### PR TITLE
Fix validators set data races

### LINKED ISSUE #145: Update default validator data races(2)
1.
```
==================
WARNING: DATA RACE
Read at 0x00c03dcd6040 by goroutine 102:
  github.com/clearmatics/autonity/consensus/tendermint/validator.(*defaultValidator).Address()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/default.go:34 +0x54
  github.com/clearmatics/autonity/consensus/tendermint/validator.(*defaultSet).GetByAddress()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/default.go:103 +0x9f
  github.com/clearmatics/autonity/consensus/tendermint/validator.(*defaultSet).IsProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/default.go:115 +0x7e
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).isProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core.go:203 +0xc2
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).logNewUnminedBlockEvent()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/unmined_block.go:102 +0x163
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).handleUnminedBlock()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/unmined_block.go:20 +0x725
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).handleEvents()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:81 +0xc10

Previous write at 0x00c03dcd6040 by goroutine 103:
  github.com/clearmatics/autonity/consensus/tendermint/validator.newDefaultSet()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/validator.go:26 +0x1c6
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Validators()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/validator.go:31 +0x15a
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).startRound()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core.go:253 +0x18fe

Goroutine 102 (running) created at:
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:17 +0x5e
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:483 +0x1ab
  github.com/clearmatics/autonity/consensus/tendermint/backend.newBlockChain()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:275 +0x4a6
  github.com/clearmatics/autonity/consensus/tendermint/backend.TestGetProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:188 +0x4f
  testing.tRunner()
      /snap/go/current/src/testing/testing.go:865 +0x163

Goroutine 103 (running) created at:
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:20 +0x9c
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:483 +0x1ab
  github.com/clearmatics/autonity/consensus/tendermint/backend.newBlockChain()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:275 +0x4a6
  github.com/clearmatics/autonity/consensus/tendermint/backend.TestGetProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:188 +0x4f
  testing.tRunner()
      /snap/go/current/src/testing/testing.go:865 +0x163
==================
```

2.
```
==================
WARNING: DATA RACE
Read at 0x00c03dcb0110 by goroutine 102:
  github.com/clearmatics/autonity/consensus/tendermint/validator.(*defaultSet).IsProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/default.go:111 +0xaa
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).isProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core.go:203 +0xc2
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).logNewUnminedBlockEvent()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/unmined_block.go:102 +0x163
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).handleUnminedBlock()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/unmined_block.go:20 +0x725
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).handleEvents()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:81 +0xc10

Previous write at 0x00c03dcb0110 by goroutine 103:
  github.com/clearmatics/autonity/consensus/tendermint/validator.newDefaultSet()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/default.go:53 +0x6b
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Validators()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/validator/validator.go:31 +0x15a
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).startRound()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core.go:253 +0x18fe

Goroutine 102 (running) created at:
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:17 +0x5e
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:483 +0x1ab
  github.com/clearmatics/autonity/consensus/tendermint/backend.newBlockChain()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:275 +0x4a6
  github.com/clearmatics/autonity/consensus/tendermint/backend.TestGetProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:188 +0x4f
  testing.tRunner()
      /snap/go/current/src/testing/testing.go:865 +0x163

Goroutine 103 (running) created at:
  github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/handler.go:20 +0x9c
  github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Start()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:483 +0x1ab
  github.com/clearmatics/autonity/consensus/tendermint/backend.newBlockChain()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:275 +0x4a6
  github.com/clearmatics/autonity/consensus/tendermint/backend.TestGetProposer()
      /mnt/sdb/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend_test.go:188 +0x4f
  testing.tRunner()
      /snap/go/current/src/testing/testing.go:865 +0x163
==================
```

### PR DESCRIPTION
**Pull request by JekaMas**
closes #145 
related to https://github.com/clearmatics/autonity/issues/141

### CODE REVIEW

File: consensus/tendermint/backend/backend.go [hunk 1]
```diff
@@ -151,10 +151,14 @@ func (sb *Backend) Broadcast(valSet tendermint.ValidatorSet, payload []byte) err
...
 		Payload: payload,
 	}
-	go sb.eventMux.Post(msg)
+	sb.postEvent(msg)
 	return nil
 }
 
+func (sb Backend) postEvent(event interface{}) {
+	go sb.eventMux.Post(event)
+}
```
Comment (aneequesafdar) [inline consensus/tendermint/backend/backend.go]: The function is still using go routines, did you mean to keep it as the commit message says you were removing them?

File: consensus/tendermint/core/handler.go [hunk 2]
```diff
@@ -142,6 +142,11 @@ func (c *core) handleEvents() {
 
 // sendEvent sends events to mux
 func (c *core) sendEvent(ev interface{}) {
+	go c.sendEventWait(ev)
+}
+
+// sendEvent sends events to mux
```
Comment (aneequesafdar) [inline consensus/tendermint/core/handler.go]: Please update the comments or remove them

File: consensus/tendermint/core/handler.go [hunk 3]
```diff
@@ -142,6 +142,11 @@ func (c *core) handleEvents() {
 
 // sendEvent sends events to mux
 func (c *core) sendEvent(ev interface{}) {
+	go c.sendEventWait(ev)
```
Comment (aneequesafdar) [inline consensus/tendermint/core/handler.go]: Same as first comment

File: consensus/tendermint/core/core.go [hunk 4]
```diff
@@ -132,7 +129,7 @@ type core struct {
 	currentHeightRoundsStates map[int64]*roundState
 
 	latestPendingUnminedBlock   *types.Block
-	latestPendingUnminedBlockMu *sync.RWMutex
```
Comment (aneequesafdar) [inline consensus/tendermint/core/core.go]: What is the reason for changing mutex pointers?

File: consensus/tendermint/backend/backend.go (hunk 1 above)
Comment (JekaMas) [reply consensus/tendermint/backend/backend.go]: Yeap, I mean to keep goroutine at the `backend`.

File: consensus/tendermint/core/handler.go (hunk 3 above)
Comment (JekaMas) [reply consensus/tendermint/core/handler.go]: I've introduced blocking and not blocking sendEvent in the `core`.

File: consensus/tendermint/core/core.go (hunk 4 above)
Comment (JekaMas) [reply consensus/tendermint/core/core.go]: It's unnecessary to have mutexes by reference when we always use `core` type by reference receiver. It was an additional complexity: to init them on type initialization.

File: consensus/tendermint/core/validator_set.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	if v.ValidatorSet == nil {
+		return false
+	}
+
+	return v.ValidatorSet.IsProposer(address)
+}
+
+func (v *validatorSet) AddValidator(address common.Address) bool {
+	v.RLock()
+	defer v.RUnlock()
```
Comment (aneequesafdar) [inline consensus/tendermint/core/validator_set.go]: Are they all meant to be acquiring Read Locks? is this intended? if so, can you please explain why

Comment (JekaMas) [reply consensus/tendermint/core/validator_set.go]: We only read the state everywhere except set method. So we need Lock only in set method and RLocks in the rest

Comment (aneequesafdar) [reply consensus/tendermint/core/validator_set.go]: I would have thought that AddValidator would be changing the state but it seems to be that you are relying on `v.ValidatorSet.AddValidator(address)` to also have a mutex, which is why I think you cannot remove the mutex from `DefaultSet` in `default.go`

**Pull request merged**